public/app/features/explore/state/epics/runRequestEpic.ts

```diff
@@ -0,0 +1,205 @@
+  datasourceId: string;
+  now: number;
+  loadingState: LoadingState;
+  series?: DataQueryResponseData[];
+  delta?: DataFrame[];
+}
+
+const processResponse = (config: ProcessResponseConfig) => {
+  const { exploreId, exploreItemState, datasourceId, now, loadingState, series, delta } = config;
+  const { queries, history } = exploreItemState;
+  const latency = Date.now() - now;
+
+  // Side-effect: Saving history in localstorage
+  const nextHistory = updateHistory(history, datasourceId, queries);
+  return [
+    historyUpdatedAction({ exploreId, history: nextHistory }),
+    processQueryResultsAction({ exploreId, latency, datasourceId, loadingState, series, delta }),
+    stateSaveAction(),
+  ];
+};
+
+interface ProcessErrorConfig {
+  exploreId: ExploreId;
+  datasourceId: string;
+  error: any;
+}
+
+const processError = (config: ProcessErrorConfig) => {
+  const { exploreId, datasourceId, error } = config;
+
+  return [processQueryErrorsAction({ exploreId, response: error, datasourceId })];
+};
+
+export const runRequestEpic: Epic<ActionOf<any>, ActionOf<any>, StoreState> = (
+  action$,
+  state$,
+  { getQueryResponse }
+) => {
+  return action$.ofType(runQueriesBatchAction.type).pipe(
+    mergeMap((action: ActionOf<RunQueriesBatchPayload>) => {
+      const { exploreId, queryOptions } = action.payload;
+      const exploreItemState = state$.value.explore[exploreId];
+      const { datasourceInstance, queries, queryIntervals, range, scanning } = exploreItemState;
+
+      // Create an observable per run queries action
+      // Within the observable create two subscriptions
+      // First subscription: 'querySubscription' subscribes to the call to query method on datasourceinstance
+      // Second subscription: 'streamSubscription' subscribes to events from the query methods observer callback
+      const observable: Observable<ActionOf<any>> = new Observable(outerObservable => {
+        outerObservable.next(queryStartAction({ exploreId }));
+
+        const datasourceId = datasourceInstance.meta.id;
+        const now = Date.now();
+        const transaction = buildQueryTransaction(queries, queryOptions, range, queryIntervals, scanning);
+        const queryState = new PanelQueryState();
+        queryState.onStreamingDataUpdated = () => {
+          const data = queryState.validateStreamsAndGetPanelData();
+          const { state, error, request, series } = data;
+          if (!data && !error) {
+            return;
+          }
+
+          if (state === LoadingState.Error) {
+            const actions = processError({ exploreId, datasourceId, error });
+            publishActions(outerObservable, actions);
+          }
+
+          if (state === LoadingState.Streaming) {
+            if (request && request.range) {
+              let newRange = request.range;
+              let absoluteRange: AbsoluteTimeRange = {
+                from: newRange.from.valueOf(),
+                to: newRange.to.valueOf(),
```

Comment:
You don't need to do any of this part... it already happened in [validateStreamsAndGetPanelData()](https://github.com/grafana/grafana/blob/master/public/app/features/dashboard/state/PanelQueryState.ts#L260)

you can just use the `request.range`